Improve performance of `az` on ubuntu hosted runners

jessehouwing opened the pull request:
# Description
Similar to: #8294

The logs show that first invocation on Ubuntu also does a command index rebuild and installs the python keyring package.

This pull request follows the same improvements that were just added to the Windows Hosted Runner.

Expected performance improvement about 10s for calling `az` for the first time.

#### Related issue:

 * Closes: https://github.com/actions/runner-images/issues/8442

## Check list
- [x] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [x] Changes are tested and related VM images are successfully generated

jessehouwing commented:
@ilia-shipitsin FYI. Image build is running in the background.

Attached logs show that Ubuntu has the same issue as windows, the performance impact is a lot lower though, but still, this should improve things further:

[dev.azure.com_jessehouwing_6484ebc3-af16-4af9-aa66-6b3398db7214__apis_build_builds_4830_logs_6.txt](https://github.com/actions/runner-images/files/12785971/dev.azure.com_jessehouwing_6484ebc3-af16-4af9-aa66-6b3398db7214__apis_build_builds_4830_logs_6.txt)

Changes are the same as on windows.

mikhailkoliada commented on images/linux/scripts/installers/azure-cli.sh:
```diff
@@ -4,10 +4,25 @@
+# Warm-up CLI
+
+# az --version performs an update check and caches the results
+az --version >> /dev/null
```
Across both scripts replace `>> /dev/null` with `> /dev/null` as we do not really append there but only redirect

jessehouwing replied:
```suggestion
az --version > /dev/null
```

mikhailkoliada commented on images/linux/scripts/installers/azure-cli.sh:
```diff
@@ -4,10 +4,25 @@
+# AZURE_EXTENSION_DIR shell variable defines where modules are installed
+# https://docs.microsoft.com/en-us/cli/azure/azure-cli-extensions-overview
+export AZURE_EXTENSION_DIR=/opt/az/azcliextensions
+echo "AZURE_EXTENSION_DIR=$AZURE_EXTENSION_DIR" | tee -a /etc/environment
```
Please use the `setEtcEnvironmentVariable`  function

jessehouwing replied:
```suggestion
setEtcEnvironmentVariable "AZURE_EXTENSION_DIR" "$AZURE_EXTENSION_DIR"
```

mikhailkoliada commented on images/linux/scripts/installers/azure-cli.sh:
```diff
@@ -4,10 +4,25 @@
+echo "AZURE_EXTENSION_DIR=$AZURE_EXTENSION_DIR" | tee -a /etc/environment
+
+export AZURE_CONFIG_DIR=/opt/az/config
+echo "AZURE_CONFIG_DIR=$AZURE_CONFIG_DIR" | tee -a /etc/environment
```
here as well :)

jessehouwing replied:
```suggestion
setEtcEnvironmentVariable "AZURE_CONFIG_DIR" "$AZURE_CONFIG_DIR"
```

mikhailkoliada commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
 ##  Desc:  Installed Azure DevOps CLI (az devops)
 ################################################################################
```
above comments can be applied for this script too I believe, not to double them :)

jessehouwing commented on images/linux/scripts/installers/azure-cli.sh:
```diff
@@ -4,10 +4,25 @@
+# az --version performs an update check and caches the results
+az --version >> /dev/null
+# az --help forces a command cache rebuild
+az --help >> /dev/null
```
```suggestion
az --help > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
+# calling az devops login to force it to install `keyring`. Login will actually fail, redirecting error to null
+echo "dummy" | az devops login >> /dev/null
+# calling az devops logout to be sure no credentials remain.
+az devops logout >> /dev/null
```
```suggestion
az devops logout > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
-invoke_tests "CLI.Tools" "Azure DevOps CLI"
\ No newline at end of file
+# calling az devops login to force it to install `keyring`. Login will actually fail, redirecting error to null
+echo "dummy" | az devops login >> /dev/null
```
```suggestion
echo "dummy" | az devops login > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
+az devops logout >> /dev/null
+
+# warm-up az devops
+az devops --help >> /dev/null
```
```suggestion
az devops --help > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
+
+# warm-up az devops
+az devops --help >> /dev/null
+az pipelines --help >> /dev/null
```
```suggestion
az pipelines --help > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
+# warm-up az devops
+az devops --help >> /dev/null
+az pipelines --help >> /dev/null
+az boards --help >> /dev/null
```
```suggestion
az boards --help > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
+az devops --help >> /dev/null
+az pipelines --help >> /dev/null
+az boards --help >> /dev/null
+az repos --help >> /dev/null
```
```suggestion
az repos --help > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
+az pipelines --help >> /dev/null
+az boards --help >> /dev/null
+az repos --help >> /dev/null
+az artifacts --help >> /dev/null
```
```suggestion
az artifacts --help > /dev/null
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
-export AZURE_EXTENSION_DIR=/opt/az/azcliextensions
-echo "AZURE_EXTENSION_DIR=$AZURE_EXTENSION_DIR" | tee -a /etc/environment
+export AZURE_DEVOPS_EXT_CONFIG_DIR=/opt/az-devops/config
+echo "AZURE_DEVOPS_EXT_CONFIG_DIR=$AZURE_DEVOPS_EXT_CONFIG_DIR" | tee -a /etc/environment
```
```suggestion
setEtcEnvironmentVariable "AZURE_DEVOPS_EXT_CONFIG_DIR" "$AZURE_DEVOPS_EXT_CONFIG_DIR"
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
+echo "AZURE_DEVOPS_EXT_CONFIG_DIR=$AZURE_DEVOPS_EXT_CONFIG_DIR" | tee -a /etc/environment
+
+export AZURE_DEVOPS_CACHE_DIR=/opt/az-devops/config/cache
+echo "AZURE_DEVOPS_CACHE_DIR=$AZURE_DEVOPS_CACHE_DIR" | tee -a /etc/environment
```
```suggestion
setEtcEnvironmentVariable "AZURE_DEVOPS_CACHE_DIR" "$AZURE_DEVOPS_CACHE_DIR"
```

mikhailkoliada commented:
@jessehouwing forgot to mention, please add the `source $HELPER_SCRIPTS/etc-environment.sh` line before the actual code, it will let you import the `setEtcEnvironmentVariable` function

jessehouwing commented on images/linux/scripts/installers/azure-cli.sh:
```diff
@@ -4,10 +4,25 @@
 ##  Desc:  Installed Azure CLI (az)
 ################################################################################
 
```
```suggestion

source $HELPER_SCRIPTS/etc-environment.sh
```

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,25 @@
 ##  Desc:  Installed Azure DevOps CLI (az devops)
 ################################################################################
 
```
```suggestion

source $HELPER_SCRIPTS/etc-environment.sh
```

mikhailkoliada commented:
@jessehouwing it looks like `az --version` does not work as it is being bitten by the permission denied error on the `/opt/az/config/` dir. I'd say we can call `sudo chmod 777 -R /opt/az` because later we do `sudo chmod 777 -R /opt` in anyway.

jessehouwing commented:
> @jessehouwing it looks like `az --version` does not work as it is being bitten by the permission denied error on the `/opt/az/config/` dir. I'd say we can call `sudo chmod 777 -R /opt/az` because later we do `sudo chmod 777 -R /opt` in anyway.

Hmz Funny... that worked on my machine...

Added the creation and setting permissions on that folder. Would it make sense to create and set permissions on that folder earlier in the process of creating the image, instead of setting the permissions trice?

Added to `az-devops` too.

jessehouwing commented:
That didn't do the trick. 😥

jessehouwing commented:
Running a build to verify changes...

jessehouwing commented:
Build succeeded.

jessehouwing commented:
/azp run ubuntu2004,ubuntu2204

jessehouwing commented:
I wonder if I can, I suppose it won't work.

jord1e commented:
@jessehouwing is something blocking you on this that I can assist in?

jord1e commented:
@mikhailkoliada could you trigger the pipeline/merge?
Thank you :)

vpolikarpov-akvelon commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,31 @@
+az artifacts --help > /dev/null
+
+invoke_tests "CLI.Tools" "Azure DevOps CLI"
+
+chown -R "$SUDO_USER:$SUDO_USER" "/opt/az"
+chown -R "$SUDO_USER:$SUDO_USER" "/opt/az-devops"
```
The same as above: I suppose it may be better to keep tests invocation at the very end of the script.

jessehouwing replied:
The tests may cause files to be generated in these directories, moving the test execution above would cause these files to have the wrong mods.

Ideally the chmod would move to the post-reboot script

jessehouwing replied:
It looks like `cleanup.sh` or `reboot.sh` would currently be the ideal locations.

vpolikarpov-akvelon replied:
Hey, we discussed it internally in our team and decided to accept this for a while. Therefore let's close this discussion.

vpolikarpov-akvelon commented on images/linux/scripts/installers/post-deployment.sh:
```diff
@@ -29,3 +29,8 @@ if npm --version
   npm cache clean --force
 fi
+
+# Clean az teletry & logs
```
```suggestion
# Clean az telemetry & logs
```

vpolikarpov-akvelon commented on images/linux/scripts/installers/post-deployment.sh:
```diff
@@ -29,3 +29,8 @@ if npm --version
 fi
+
+# Clean az teletry & logs
+rm -rf /opt/az/config/telemetry.txt
+rm -rf /opt/az/config/telemetry
+rm -rf /opt/az/config/logs
```
Please don't use `-r` flag when delete files, keep it only for directories.

jessehouwing commented on images/linux/scripts/installers/azure-cli.sh:
```diff
@@ -4,10 +4,30 @@
+az --help > /dev/null
+
 invoke_tests "CLI.Tools" "Azure CLI"
+
+chown -R "$SUDO_USER:$SUDO_USER" "/opt/az"
```
The tests may cause files to be generated in these directories, moving the test execution above would cause these files to have the wrong mods. 

Ideally the chmod would move to the post-reboot script

jessehouwing replied:
It looks like `cleanup.sh` or `reboot.sh` would currently be the ideal locations.

vpolikarpov-akvelon replied:
Hey, we discussed it internally in our team and decided to accept this for a while. Therefore let's close this discussion.

jessehouwing commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,31 @@
 
-invoke_tests "CLI.Tools" "Azure DevOps CLI"
\ No newline at end of file
+# calling az devops login to force it to install `keyring`. Login will actually fail, redirecting error to null
+echo "dummy" | az devops login > /dev/null
+# calling az devops logout to be sure no credentials remain.
+az devops logout > /dev/null
```
if the script returns an error, then I'd guess the installation of the python credential package could have failed. I'd say the script should break in that case...

vpolikarpov-akvelon replied:
Doesn't fake authentication fail regardless of package installation success?

jessehouwing replied:
It succeeds regardless of what kind of PAT you feed it. 

It's the logout that may complain a logout wasn't needed.

jessehouwing commented:
@vpolikarpov-akvelon Fixed a couple and replied to remaining comments.

vpolikarpov-akvelon commented on images/linux/scripts/installers/azure-devops-cli.sh:
```diff
@@ -4,12 +4,31 @@
-echo "AZURE_EXTENSION_DIR=$AZURE_EXTENSION_DIR" | tee -a /etc/environment
+source $HELPER_SCRIPTS/etc-environment.sh
+export AZURE_DEVOPS_EXT_CONFIG_DIR=/opt/az-devops/config
+setEtcEnvironmentVariable "AZURE_DEVOPS_EXT_CONFIG_DIR" "$AZURE_DEVOPS_EXT_CONFIG_DIR"
```
Could you please create PR to fix this variable for Windows as well?

jessehouwing replied:
Sure. I shall also submit one to clean up the telemetry on windows.

jessehouwing commented:
As requested:

- https://github.com/actions/runner-images/pull/8800

jessehouwing commented:
@mikhailkoliada 👍

jessehouwing commented on images/ubuntu/scripts/build/install-azure-cli.sh:
```diff
@@ -4,10 +4,30 @@
+# AZURE_EXTENSION_DIR shell variable defines where modules are installed
+# https://docs.microsoft.com/en-us/cli/azure/azure-cli-extensions-overview
+export AZURE_EXTENSION_DIR=/opt/az/azcliextensions
+setEtcEnvironmentVariable "AZURE_EXTENSION_DIR" "$AZURE_EXTENSION_DIR"
```
This PR seems to have changed the convention away from the `setEtcEnvironmentVariable` call back to `| tee -a /etc/environment`. I'd be happy to change these calls either way.

jessehouwing replied:
@mikhailkoliada your thoughts? I'd like to get this closed with the pace at which you folks are updating files, I need to rebase this PR every other day at the moment ;)>

jessehouwing commented on images/ubuntu/scripts/build/install-azure-devops-cli.sh:
```diff
@@ -9,7 +9,30 @@
 export AZURE_EXTENSION_DIR=/opt/az/azcliextensions
 echo "AZURE_EXTENSION_DIR=$AZURE_EXTENSION_DIR" | tee -a /etc/environment
```
These lines aren't needed, but were introduced by the refactoring of the build scripts.

The environment variable is already set in `install-azure-cli.ch` and it makes more sense there, as it's a config setting of `az-cli` and not of `az-devops`.

jessehouwing commented:
Ping @mikhailkoliada @jord1e  @vpolikarpov-akvelon I really want this thing done. What can we do to move it off the list? I've rebased since the restructure, it looks like some styling things changed, hence my open questions above.

I'd like to resolve those ASAP and then close this, it is taking longer than I'd anticipated.

mikhailkoliada commented:
After observations I do not think we are ready to change this aspect.

1. I am not sure that customers won't be broken if we change the configuration directory and make it the global one
2. We do not observe hight demand in az optimisation from customers in general
3. Even if (2) were true I do not think it should be fixed from az upstream side and not the runner-images side

jord1e commented:
My initial `az login`s take 22+ seconds first run, which is insane.

> 1. I am not sure that customers won't be broken if we change the configuration directory and make it the global one
How would it break customers' setups? All configuration is handled via https://learn.microsoft.com/en-us/cli/azure/devops?view=azure-cli-latest#az-devops-configure for az devops, is it not? I get the concern but all configuration should be done via the CLI anyways.
> 2. We do not observe hight demand in az optimisation from customers in general
Because they don't know about these things in general as it would only be observed in CI environments where there is no command & update index caching
> 3. Even if (2) were true I do not think it should be fixed from az upstream side and not the runner-images side
How would az upstream be able to create an index and do an update check before the executable was every called?

And it is implemented on the windows runners already, why would implementing it on Ubuntu change anything summed up in these statements?
https://github.com/actions/runner-images/pull/8427/files

jessehouwing commented:
After observations I do not think we are ready to change this aspect.

 1. I am not sure that customers won't be broken if we change the configuration directory and make it the global one

> This isn't as much of a change, as much as it is formalizing the default configuration. If not instructed through the environment variable, `az` will create a global configuration on first invocation and use that. If people do not want to use the default configuration, they should be overriding the environment variables to begin with, and that would still be possible.

 2. We do not observe high demand in `az` optimisation from customers in general

> I do see high demand for faster runners. Many organizations spend a large amount of time trying to speed up, parallelize and optimize their workflows. That there is no demand specifically for `az` might be because people aren't aware a 15s gain is available.

 3. Even if (2) were true I do not think it should be fixed from az upstream side and not the runner-images side

> While I agree the `az` and `az devops` team could do more here, they're not building the tool to be primarily used on a hosted runner. `az` seems to me first and foremost a developer/administrator tool. The default setup works pretty well in those circumstances. The ephemeral hosted runner is a pretty specific environment and the installer isn't optimizing for that at the moment.

I think it would be possible on the ubuntu image to forego setting the environment variables, still have a reasonable speed improvement. The reason I set the environment variables on the ubuntu runner, is to make the setup aligned to the windows setup, where (due to the removal of the provisioning account) it was deemed preferred to set these environment variables.

Note that the current scripts already set a variable for the Azure DevOps extension installation folder.
